### PR TITLE
String Func: repeat/substring/substr/mid

### PR DESCRIPTION
**Pull request by Veeupup**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

string func:
- [x] repeat
- [x] substring
- [x] mid
- [x] substr  
- [x] substring_index

## Changelog

- Improvement

## Related Issues

Fixes #issue

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/functions/src/scalars/strings/substring.rs [hunk 1]
```diff
@@ -14,235 +14,97 @@
...
-        let dt = DataType::String;
-        Ok(DataTypeAndNullable::create(&dt, nullable))
+
+        Ok(StringType::arc())
     }
 
-    fn eval(&self, columns: &DataColumnsWithField, input_rows: usize) -> Result<DataColumn> {
-        let s_column = columns[0].column().cast_with_type(&DataType::String)?;
-        let p_column = columns[1].column().cast_with_type(&DataType::Int64)?;
+    fn eval(&self, columns: &ColumnsWithField, input_rows: usize) -> Result<ColumnRef> {
```
Comment (Veeupup) [inline common/functions/src/scalars/strings/substring.rs:80]: thanks for column viewer, this complicated match could be avoid cc @sundy-li

**Review by PsiACE — APPROVED** [review]

File: common/functions/src/scalars/strings/substring.rs (hunk 1 above)
Comment (sundy-li) [reply common/functions/src/scalars/strings/substring.rs:80]: Great!  But I suggest merging main into this pr to use the latest version of `Viewer `.

Comment (sundy-li) [reply common/functions/src/scalars/strings/substring.rs:80]: let viewer = i16::try_create_viewer(&column)?;

    assert_eq!(viewer.size(), 1024);
    assert!(!viewer.null_at(0));
    assert!(!viewer.null_at(99));

    for (idx, value) in viewer.iter().enumerate() {
        if viewer.null_at(idx) {
            assert_eq!(value, 0);
        } else {
            assert_eq!(value, 123i16);
        }
    }

**Comment by sundy-li** [comment]
@mergify update

File: common/functions/src/scalars/strings/substring.rs (hunk 1 above)
Comment (Veeupup) [reply common/functions/src/scalars/strings/substring.rs:81]: resolved

**Comment by sundy-li** [comment]
Good job, almost -300  LOC

**Pull request merged**